geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/CommandInitializer.java

```diff
@@ -111,37 +89,76 @@
  * @since GemFire 5.7
  */
 
-public class CommandInitializer {
+public class CommandInitializer implements CommandRegistry {
 
-  @Immutable
-  static final Map<KnownVersion, Map<Integer, Command>> ALL_COMMANDS = initializeAllCommands();
+  @Deprecated
+  @MakeNotStatic
+  static final CommandInitializer instance = new CommandInitializer();
 
   /**
-   * Register a new command with the system.
+   * Gets legacy singleton instance.
+   *
+   * @deprecated Efforts should be made to get and instance from the cache or other object.
    *
-   * @param messageType - An ordinal for this message. This must be something defined in MessageType
-   *        that has not already been allocated to a different command.
-   * @param versionToNewCommand The command to register, for different versions. The key is the
-   *        earliest version for which this command class is valid (starting with GFE_57). The value
-   *        is the command object for clients starting with that version.
+   * @return legacy singleton instance. Instance is not immutable.
    */
-  public static void registerCommand(int messageType,
+  @Deprecated
+  public static CommandInitializer getDefaultInstance() {
+    return instance;
+  }
+
+  final Map<KnownVersion, Map<Integer, Command>> unmodifiableRegisteredCommands;
+  final LinkedHashMap<KnownVersion, ConcurrentMap<Integer, Command>> modifiableRegisteredCommands;
```

Comment:
Why must we retain a reference to the modifiable commands? It looks like it is to support `CqServiceFactoryImpl` calling `register()`. Why do we keep the two separate maps?

Reply:
Ah... to maintain the contract that the maps returned by the `get(Version)` method are immutable views. The previous implementation made the outer map immutable but not the inner map so the caller could technically mutate the map causing who knows what issues. Now the only way to mutate is through `register`, which mutates the backing maps but keeps the immutable wrappers intact.

Reply:
You can think of the unmodifiable version of the map as just a map of unmodifiable wrappers. It doesn't contain any data itself. I am not committed to it though if you don't like it.

Reply:
The unmodifiable maps are just wrappers and not copies so changes made to the modifiable map are visible to unmodifiable map.

Reply:
Thanks @pivotal-jbarrett.